src/graphics/env-lighting.js

```diff
@@ -0,0 +1,397 @@
+    // device.setRenderTarget(oldRt);
+    // device.updateBegin();
+};
+
+// generate mipmaps for the given target texture
+// target is either 2d equirect or cubemap with mipmaps = false
+const generateMipmaps = (target) => {
+    const device = target.device;
+    const gl = device.gl;
+    const numLevels = calcLevels(target.width, target.height);
+
+    let prevLevel = null;
+    for (let i = 1; i < numLevels; ++i) {
+        const level = new Texture(device, {
+            name: target.name + '-tmp-' + i,
+            cubemap: target.cubemap,
+            width: Math.max(1, Math.floor(target.width >> i)),
+            height: Math.max(1, Math.floor(target.height >> i)),
+            format: target.format,
+            type: target.type,
+            addressU: target.addressU,
+            addressV: target.addressV,
+            fixCubemapSeams: target.fixCubemapSeams,
+            mipmaps: false
+        });
+
+        // downsample texture using single sample reproject
+        reprojectTexture(prevLevel || target, level, {
+            numSamples: 1
+        });
+
+        // copy into target's mipmap
+        copyMipmap(target, level);
+
+        if (prevLevel) {
+            prevLevel.destroy();
+        }
+        prevLevel = level;
+    }
+
+    if (prevLevel) {
+        prevLevel.destroy();
+    }
+
+    // set filtering while bypassing the engine
+    device.setTexture(target, 0);
+    gl.texParameteri(target._glTarget, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
+    gl.texParameteri(target._glTarget, gl.TEXTURE_MIN_FILTER, gl.LINEAR_MIPMAP_LINEAR);
+    target._mipmaps = true;
+};
+
+// helper functions to support prefiltering lighting data
+class EnvLighting {
+    /**
+     * @private
+     * @function
+     * @name generateSkyboxCubemap
+     * @description Generate a skybox cubemap in the correct pixel format from the source texture.
+     * @param {Texture} source - The source texture. This is either a 2d texture in equirect format or a cubemap.
+     * @param {number} size - Size of the resulting texture. Specify 0 for automatic sizing.
+     * @returns {Texture} The resulting cubemap.
+     */
+    static generateSkyboxCubemap(source, size = 0) {
+        if (size === 0) {
+            size = source.cubemap ? source.width : source.width / 4;
+        }
+
+        const device = source.device;
+
+        // generate faces cubemap
+        const result = new Texture(device, {
+            name: 'skyboxFaces',
+            cubemap: true,
+            width: size,
+            height: size,
+            type: TEXTURETYPE_RGBM,
+            addressU: ADDRESS_CLAMP_TO_EDGE,
+            addressV: ADDRESS_CLAMP_TO_EDGE,
+            fixCubemapSeams: fixCubemapSeams,
+            mipmaps: false
+        });
+
+        reprojectTexture(source, result, {
+            numSamples: 1024
+        });
+
+        return result;
+    }
+
+    /**
+     * @private
+     * @function
+     * @name generateLightingSource
```

Comment:
perhaps this could do with a better name .. we pass in source (parameter) and it generates lighting source. Those are too similar?

Reply:
This function generates the texture that'll be used for generating the reflection and ambient maps.

The texture must have mipmaps and have a filterable pixel format.

Any idea of a better name?

Reply:
not sure .. I'm warming up to this name already :)